packages/nuxt/src/core/nuxt.ts

```diff
@@ -124,18 +124,28 @@ async function initNuxt (nuxt: Nuxt) {
   await nuxt.callHook('modules:before')
   const modulesToInstall = [...nuxt.options.modules]
 
+  const watchedPaths = new Set<string>()
+
   // Automatically register user modules
   for (const config of nuxt.options._layers.map(layer => layer.config)) {
     const userModules = await resolveFiles(config.srcDir, [
       `${config.dir?.modules || 'modules'}/*{${nuxt.options.extensions.join(',')}}`,
       `${config.dir?.modules || 'modules'}/*/index{${nuxt.options.extensions.join(',')}}`
     ])
-    modulesToInstall.push(...userModules)
+    for (const mod of userModules) {
+      modulesToInstall.push(mod)
+      watchedPaths.add(relative(config.srcDir, mod))
+    }
   }
 
   // Register ad-hoc modules
   modulesToInstall.push(...nuxt.options._modules)
 
+  nuxt.hook('builder:watch', (event, path) => {
```

Comment:
Since this is a hard restart request, we might use `hookOnce` to avoid doing multiple requests:

```suggestion
  nuxt.hookOnce('builder:watch', (event, path) => {
```